pyiron_base/jobs/flex/executablecontainer.py

```diff
@@ -72,22 +79,72 @@ def set_job_type(
         if default_input_dict is not None:
             self.input.update(default_input_dict)
 
-    def write_input(self):
-        if self._write_input_funct is not None:
-            self._write_input_funct(
-                input_dict=self.input.to_builtin(),
-                working_directory=self.working_directory,
-            )
+    def generate_calculate_function_kwargs(self) -> dict:
+        """
+        Generate keyword arguments for the calculate() function
+
+        Returns:
+            dict: keyword arguments for the calculate() function
+        """
+        return {
+            "input_dict": self.input.to_builtin(),
+            "executable_dict": {
+                "executable": self.executable.executable_path,
+                "shell": True,
+                "working_directory": self.working_directory,
+                "conda_environment_name": self.server.conda_environment_name,
+                "conda_environment_path": self.server.conda_environment_path,
+            },
+        }
+
+    def generate_calculate_function(self) -> callable:
```

Comment:
Can be inlined at the only call site, unless you see this being called elsewhere in the future.

Reply:
I want to provide the user with a simple way to convert a pyiron job object to a python function call. At the current stage the user can create the job and interact with it and then call `generate_calculate_function()` and `generate_calculate_function_kwargs()` to have the corresponding `calculate()` function which executes the calculation of the specific job. So yes, I consider the user to call the functions externally, still I am not sure if it might make more sense to rename them or have them in a combined function.

Reply:
If it's user facing, than it's ok to have it.  But then it needs docstring that explains how to use it, like your example in the [meeting notes](https://github.com/orgs/pyiron/discussions/231#discussioncomment-9751091). 

Couldn't `generate_calculate_function` return a closure that already has the arguments baked in?  It's a bit awkward to get the function and the kwargs separately when you are really expected to call and use them together.

Reply:
At the moment this is correct, but at the same time I would like to have the option to submit these functions to an `concurrent.futures.Executor` so I guess we need the function and the keyword arguments separated. Still I am open to have both merged into one function.

Reply:
I've put together a small snippet that I mention in the meeting.  It defines a picklable `Caller` object that would hold onto the write and collect function as well as their arguments.  I think it clears up the logic quite a bit and would also immediately be able to be used with the executors (assuming the passed write and collect function are (cloud-)picklable, which should be no issue at all.

```python
import os

from pickle import loads, dumps

def write(input_dict):
    print("Im writing to disk!")
    print(input_dict)

def collect():
    print("I'm reading this from disk, honest!")
    return 42

def myfunc(f0, f1, *args, **kwargs):
    print(args)
    r = f0(kwargs)
    return f1()

class Caller:
    __slots__ = ('func', 'args', 'kwargs')

    def __init__(self, func, *args, **kwargs):
        self.func = func
        self.args = args
        self.kwargs = kwargs

    def __call__(self):
        return self.func(*self.args, **self.kwargs)

def calculate(write_function, collect_function, input_dict=None, executable_dict=None):
    if input_dict is not None:
        write_function(input_dict)

    # do something with executable_dict
    os.system("date")

    output = collect_function()
    return output


def generate_calculate(job, write=None, collect=None):
    return Caller(
            calculate,
            # presumably write/collect would be defined per job/simulation code
            # they could live as static methods on the job classes themselves,
            # then we'd be able to use them here as job.write, job.collect
            write,
            collect,
            **job.generate_calculate_function_kwargs()
    )

if __name__ == "__main__":
    loads(dumps(Caller(myfunc, write, collect, 3, [3,5], a='More kwargs')))()
```

Reply:
Ah, no, I was not clear enough.  This `Caller(...)` does not execute the function, you need to call the resulting object.

```python
with ProcessPoolExecutor() as exe:
    call = Caller(myfunc, write, collect, 3, [3,5], a = 'More kwargs')
    fs = exe.submit(call)
    print(fs.result())
```

This works for me with the plain `ProcessPoolExecutor` and `pympipool.Executor` when the functions are defined in a file.  When defined in a notebook it has to fail with the normal pickler of course.  I could only test `pympipool.Executor` locally, because on a notebook on the cluster it seems to submit automatically to SLURM(?) and my queue is already full.

In practice my understanding was that `write_input_function` and `collect_function` would be

Reply:
For `pympipool` you can enforce a specific backend by setting `backend="local"` or basically anything other than `"flux", "slurm", "auto"`. The current default is `"auto"` which causes pympipool to use SLURM when available.

Reply:
Here's an updated example that works on my local machine.  It's not running on the cluster, because we don't seem to have pympipool==0.8 available yet.

```python
import os

from concurrent.futures import ProcessPoolExecutor
from pympipool import Executor
from pickle import loads, dumps

def write(input_dict):
    print("Im writing to disk!")
    print(input_dict)

def collect():
    print("I'm reading this from disk, honest!")
    return 42

def myfunc(f0, f1, *args, **kwargs):
    print(args)
    r = f0(kwargs)
    return f1()

class Caller:
    __slots__ = ('func', 'args', 'kwargs')

    def __init__(self, func, *args, **kwargs):
        self.func = func
        self.args = args
        self.kwargs = kwargs

    def __call__(self):
        return self.func(*self.args, **self.kwargs)

def calculate(write_function, collect_function, input_dict=None, executable_dict=None):
    if input_dict is not None:
        write_function(input_dict)

    # do something with executable_dict
    os.system("date")

    output = collect_function()
    return output


def generate_calculate(job, write=None, collect=None):
    return Caller(
            calculate,
            # presumably write/collect would be defined per job/simulation code
            # they could live as static methods on the job classes themselves,
            # then we'd be able to use them here as job.write, job.collect
            write,
            collect,
            **job.generate_calculate_function_kwargs()
    )

if __name__ == "__main__":
    call = Caller(myfunc, write, collect, 3, [3,5], a='More kwargs')
    loads(dumps(call))()
    with ProcessPoolExecutor() as exe:
        fs = exe.submit(call)
        print(fs.result())
    with Executor() as exe:
        fs = exe.submit(call)
        print(fs.result())
```

Reply:
I implemented the `Caller` class and it is now working the regular executor, at least for standard jobs like LAMMPS.